flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StopMode.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+
+/**
+ * Tells if the job is stopping because of consuming all data. {@link #DRAINED} means the job is
+ * stopping either with a stop-with-savepoint --drain or due to consuming all records in the source.
+ * A drained pipeline should call {@link StreamOperator#finish()} on all operators.
+ */
+@Internal
+public enum StopMode {
```

Comment:
Hi Dawid do you think we might keep using `boolean shouldDrain`? I'm asking since currently it seems we have a mix use of the enum and the boolean, and have transformation on received `CheckpointBarrier` or emit `EndOfData`. But I'm might not be very strongly argue for that.

Reply:
The reason I went for `StopMode` instead of a `boolean` is that I learned using enums is more descriptive. I agree it is a pity I use a boolean flag in the `notifyEndOfData`, but I did it because it is in a different module. I'd prefer using enum there as well.

I am also not strong on this, but I am slightly leaning towards usage of the `StopMode`, just for the sake of it being more descriptive.

Reply:
I am not the biggest fan of putting it in `flink-core` it is a rather user facing module and the `StopMode` is an entirely internal flag :(

Moving it to `flink-runtime` is also not perfect because it strengthens the wrong dependency order (flink-streaming should not depend on runtime), but I think it would be a lesser evil.